### PR TITLE
Expand Permalink, split into GeneratorParameters

### LINKED ISSUE #2596: Permalinks: Report error message when it's for a different Randovania version.

### PR DESCRIPTION
**Pull request by henriquegemignani**
Fixes #2596.

### CODE REVIEW

File: randovania/gui/game_session_window.py [hunk 1]
```diff
@@ -1032,14 +1033,15 @@ async def import_permalink(self):
             return
 
         permalink = dialog.get_permalink_from_field()
-        if permalink.player_count != self._game_session.num_rows:
+        parameters = permalink.parameters
+        if parameters.player_count != self._game_session.num_rows:
             return await async_dialog.warning(
                 self, "Incompatible permalink",
-                f"Given permalink is for {permalink.player_count} players, but "
+                f"Given permalink is for {parameters} players, but "
```
Comment (duncathan) [inline randovania/gui/game_session_window.py:1040]: `parameters.player_count`

File: randovania/gui/main_window.py [hunk 2]
```diff
@@ -279,19 +286,32 @@ def _open_faq(self):
...
             layout = simplified_patcher.generate_layout(progress_update=progress_update,
-                                                        permalink=permalink,
+                                                        parameters=permalink.parameters,
                                                         options=self._options)
             progress_update(f"Success! (Seed hash: {layout.shareable_hash})", 1)
             return layout
 
         new_layout = await BackgroundProcessDialog.open_for_background_task(work, "Creating a game...")
+
+        if permalink.seed_hash is not None and permalink.seed_hash == new_layout.shareable_hash_bytes:
```
Comment (duncathan) [inline randovania/gui/main_window.py:303]: should this be `!=`?

File: randovania/layout/generator_parameters.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                yield from bitpacking.pack_array_element(preset, previous_unique_presets)
+            else:
+                previous_unique_presets.append(preset)
+                yield from preset.bit_pack_encode({"manager": manager})
+
+        games = []
+        for preset in self.presets.values():
+            if preset.game not in games:
+                games.append(preset.game)
+                yield _game_db_hash(preset.game), 256
```
Comment (duncathan) [inline randovania/layout/generator_parameters.py:55]: set comprehension might be cleaner:

```py
games = {preset.game for preset in self.presets.values()}
for game in games:
    yield _game_db_hash(game), 256
```

File: randovania/layout/generator_parameters.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                preset = decoder.decode_element(previous_unique_presets)
+            else:
+                preset = Preset.bit_pack_unpack(decoder, {"manager": manager})
+                previous_unique_presets.append(preset)
+            presets[index] = preset
+
+        games = []
+        for preset in presets.values():
+            if preset.game not in games:
+                games.append(preset.game)
```
Comment (duncathan) [inline randovania/layout/generator_parameters.py:79]: similar to above, you could use a set comprehension

File: randovania/layout/permalink.py [hunk 5]
```diff
@@ -40,140 +50,132 @@ def rotate_bytes(data: Iterable[int], rotation: int, per_byte_adjustment: int,
...
+        )
 
     @property
     def as_base64_str(self) -> str:
         try:
-            return base64.b64encode(self.as_bytes, altchars=b'-_').decode("utf-8")
+            encoded = PermalinkBinary.build({
+                "value": {
+                    "header": {
+                        "is_dev_version": True,
```
Comment (duncathan) [inline randovania/layout/permalink.py:133]: why hardcoded to true?

File: randovania/layout/permalink.py [hunk 6]
```diff
@@ -40,140 +50,132 @@ def rotate_bytes(data: Iterable[int], rotation: int, per_byte_adjustment: int,
...
+            raise ValueError("Extra text at the end")
 
-    @property
-    def player_count(self) -> int:
-        return len(self.presets)
+        except construct.core.StreamError:
+            raise ValueError("Missing text at the end")
+
+        except construct.core.ChecksumError:
+            raise ValueError("Incorrect checkum")
```
Comment (duncathan) [inline randovania/layout/permalink.py:164]: checksum

**Review by duncathan — CHANGES_REQUESTED** [review]

File: randovania/gui/main_window.py (hunk 2 above)
Comment (henriquegemignani) [reply randovania/gui/main_window.py:303]: ooops my test

File: randovania/layout/generator_parameters.py (hunk 3 above)
Comment (henriquegemignani) [reply randovania/layout/generator_parameters.py:55]: not deterministic

File: randovania/layout/permalink.py (hunk 5 above)
Comment (henriquegemignani) [reply randovania/layout/permalink.py:133]: no longer a thing, good catch

File: randovania/layout/generator_parameters.py (hunk 3 above)
Comment (duncathan) [reply randovania/layout/generator_parameters.py:55]: could always just sort it, but yeah makes sense

**Pull request merged**